src/usrp_driver.py

```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env python3
+
+"""
+    usrp_driver
+    ~~~~~~~~~~~
+    A python wrapper which launches the usrp_driver, captures logs into structlog,
+    and enables python controllers and managers to interact with the drivers.
+
+    :copyright: 2023 SuperDARN Canada
+    :author: Adam Lozinsky
+"""
+import os
+import subprocess
+import faulthandler
+import argparse as ap
+import pty
+import re
+import time
+
+
+def main():
+    """
+    Given the parsed arguments this runs the USRP driver in a subproccess.
+    The subprocess stdout and stderr are streamed through a pty to be read in realtime,
+    parsed, then logged using structlog.
+    """
+
+    # TODO (Adam): - console should not clear on crash
+    #              - debug mode does not work
+    faulthandler.enable()
+    parser = ap.ArgumentParser(description='Wrapper to the USRP driver')
+    parser.add_argument("run_mode", help="The mode to run, switches scons builds and some arguments to "
+                                         "modules based on this mode. Commonly 'release'.")
+    parser.add_argument('--c_debug_opts', help='A C debug run options string',
+                        default='')
+    args = parser.parse_args()
+
+    path = os.environ["BOREALISPATH"]
+    cmd = f"source {path}/mode {args.run_mode}; {args.c_debug_opts} usrp_driver"
+    log.debug('usrp_driver start command', command=cmd)
+
+    # If you are here to work on the code below this comment I bid you good luck!
+    # This will only work on LINUX!
+
+    # First we open a pseudoterminal and get the file descriptors for the stdout stream
+    stdout_master_fd, stdout_slave_fd = pty.openpty()
+
+    # Then start usrp_driver executable via subprocess.Popen
+    # We need text=False and universal_newlines=False to ensure byte mode (default)
+    # We need shell=True, even though it is less secure so that we can 'source mode'
+    # We use close_fds=True so that the process does not hang open printing empty log lines
+    # We need to send stderr to stdout to have one stream, UHD sends logs to stderr, we do stdout
+    driver = subprocess.Popen([cmd],
```

Comment:
Is it possible to set stdin in the Popen call in order to function in `engineeringdebug` mode? Then we could could check if `__debug__` is set in order to determine whether we need to send messages to the subprocess or not.

Reply:
This has been fixed.